### PR TITLE
Enable user authentication by access token

### PR DESCRIPTION
**Pull request by jhkchan**
As the case written in http://stackoverflow.com/questions/4623974/design-for-facebook-authentication-in-an-ios-app-that-also-accesses-a-secured-we, it is possible that access tokens obtained from mobile apps (iOS or Android) would be used for server side authentication.

This change set retrieves access_token pass by request during facebook login check to enable facebook user authentication by access token. This also handles the issue #210.

### CODE REVIEW

**Comment by diegoholiveira** [comment]
Using the access_token for auth can cause security issues? I mean, if someone get your access_token it will be possible to pass as you, right?

**Comment by jhkchan** [comment]
I agree. So developers should understand that it should be passed by HTTPS only just like using graph.facebook.com. This is a valid use-case.

**Comment by diegoholiveira** [comment]
Can you provide some documentation about it? If you include examples of how and where to use it will be awesome.

**Comment by jhkchan** [comment]
Should I do it here, or edit README.mdand pull again? I will provide detailed steps on API calls using Facebook SDK and called by mobile app.

**Comment by diegoholiveira** [comment]
Do it in the README.md and pull again, please. :)

**Comment by jhkchan** [comment]
Is this readme okay?

**Comment by diegoholiveira** [comment]
It's fine.

I'll merge it once the tests are green. Take a look here: https://travis-ci.org/FriendsOfSymfony/FOSFacebookBundle/jobs/4073267

**Comment by jhkchan** [comment]
I read it but don't understand why updating readme will make build fail? I saw the signature of AuthenticationException is different but I didn't change there. Is there something that I need to fix?

File: Security/Authentication/Provider/FacebookProvider.php [hunk 1]
```diff
@@ -69,6 +69,9 @@ public function authenticate(TokenInterface $token)
         }
 
         try {
+            if (!is_null($token->getAccessToken())) {
+                $this->facebook->setAccessToken($token->getAccessToken());
```
Comment (diegoholiveira) [inline Security/Authentication/Provider/FacebookProvider.php:73]: @jhkchan probably this is the cause of failure on the tests. You should verify and treat any exception raised by this call.

Comment (jhkchan) [inline Security/Authentication/Provider/FacebookProvider.php:73]: @diegoholiveira This line throws no exception. I am still investigating the cause of the failed test. thanks.

**Comment by jhkchan** [comment]
I passed the unit test on my local side, but not on Travis. I find Travis uses Symfony 2.2, not Symfony 2.1. That could be the problem.

**Comment by diegoholiveira** [comment]
I'll tag specific versions for symfony 2.1 and symfony 2.2 (I guess I can do it later today).

**Comment by jhkchan** [comment]
I have tried to reproduce the issue and identified the problem exists in the current codebase.

To illustrate, I would like to reproduce the error by the following reduced example:

``` php
<?php

try {
    throw new \Exception('message');
} catch (\Exception $failed) {
    throw new \RuntimeException($failed->getMessage(), null, (int)$failed->getCode(), $failed);
}

?>
```

This will return 

```
PHP Fatal error:  Wrong parameters for Exception([string $exception [, long $code [, Exception $previous = NULL]]]) in
XXX/xxx.php on line 6
```

The correct code would be removing the argument `null`.

References:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Exception/AuthenticationException.php
http://php.net/manual/en/class.runtimeexception.php
http://php.net/manual/en/class.exception.php

Why the current codebase would not trigger this because all are caught by the first catch identifying `AuthenticationException`. I am not sure why my code triggered the coverage path to `\Exception` but that is the case. Should I commit the fix to here as well? Thanks!

**Comment by diegoholiveira** [comment]
Could you update your PR with the last version on master? This last version use a stable version of symfony 2.1 and maybe it can solve this problem. If not, I'll take a look to see what's happen.

thanks.

**Comment by jhkchan** [comment]
Please let me know what I can do. Thanks.

**Comment by diegoholiveira** [comment]
I'll take a look on it and give you an answer.

**Comment by jhkchan** [comment]
I would suggest you to rebuild again on Travis. Something was wrong. Now I have mine passed: https://travis-ci.org/jhkchan/FOSFacebookBundle

**Comment by diegoholiveira** [comment]
I will take a look on your branch on my machine, but I need to take some
time for do it :(

**Pull request merged**